Mathlib/Analysis/Complex/ValueDistribution/CountingFunction.lean

```diff
@@ -0,0 +1,217 @@
+locally finite support to maps `ℝ → ℝ`.  Given `D`, the result map `logCounting D` takes a real `r :
+ℝ` to a logarithmically weighted measure of values that `D` takes within the disk `∣z∣ ≤ r`.
+
+Implementation Note: In case where `z = 0`, the term `log (r * ‖z‖⁻¹)` evaluates to zero, which is
+typically different from `log r - log ‖z‖ = log r`. The summand `(D`0) * log r` compensates this,
+producing cleaner formulas then the logarithmic counting function is used in the main theorms of
+Value Distribution Theory.  We refer the reader to page 164 of [Lang: Introduction to Complex
+Hyperbolic Spaces](https://link.springer.com/book/10.1007/978-1-4757-1945-1) for more details.
+-/
+noncomputable def logCounting {E : Type*} [NormedAddCommGroup E] [ProperSpace E] :
+    locallyFinsuppWithin (univ : Set E) ℤ →+ (ℝ → ℝ) where
+  toFun D := fun r ↦ ∑ᶠ z, D.toClosedBall r z * log (r * ‖z‖⁻¹) + (D 0) * log r
+  map_zero' := by aesop
+  map_add' D₁ D₂ := by
+    simp only [Set.top_eq_univ, map_add, coe_add, Pi.add_apply, Int.cast_add]
+    ext r
+    have {A B C D : ℝ} : A + B + (C + D) = A + C + (B + D) := by ring
+    rw [Pi.add_apply, this]
+    congr 1
+    · have h₁s : ((D₁.toClosedBall r).support ∪ (D₂.toClosedBall r).support).Finite := by
+        apply Set.finite_union.2
+        constructor
+        <;> apply finiteSupport _ (isCompact_closedBall 0 |r|)
+      repeat
+        rw [finsum_eq_sum_of_support_subset (s := h₁s.toFinset)]
+        try simp_rw [← Finset.sum_add_distrib, ← add_mul]
+      repeat
+        intro x hx
+        by_contra
+        simp_all
+    · ring
+
+/--
+Evaluation of the logarithmic counting function at zero yields zero.
+-/
+@[simp] lemma logCounting_eval_zero {E : Type*} [NormedAddCommGroup E] [ProperSpace E]
+    (D : locallyFinsuppWithin (univ : Set E) ℤ) :
+    logCounting D 0 = 0 := by
+  rw [logCounting]
+  simp
+
+end Function.locallyFinsuppWithin
+
+/-!
+## The Logarithmic Counting Function of a Meromorphic Function
+-/
+
+namespace ValueDistribution
+
+variable
+  {𝕜 : Type*} [NontriviallyNormedField 𝕜] [ProperSpace 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] [CompleteSpace E]
+  {U : Set 𝕜} {f g : 𝕜 → E} {a : WithTop E} {a₀ : E}
+
+variable (f a) in
+/--
+The logarithmic counting function of a meromorphic function.
+
+If `f : 𝕜 → E` is meromorphic and `a : WithTop E` is any value, this is a logarithmically weighted
+measure of the number of times the function `f` takes a given value `a` within the disk `∣z∣ ≤ r`,
+counting multiplicities.  In the special case where `a = ⊤`, it counts the poles of `f`.
+-/
+noncomputable def logCounting : ℝ → ℝ := by
+  by_cases h : a = ⊤
+  · exact (divisor f univ)⁻.logCounting
+  · exact (divisor (fun z ↦ f z - a.untop₀) univ)⁺.logCounting
+
+/--
+For finite values `a₀`, the logarithmic counting function `logCounting f a₀` is the counting
+function associated with the zero-divisor of the meromorphic function `f - a₀`.
+-/
+lemma logCounting_finite :
```

Comment:
```suggestion
lemma logCounting_coe :
```
For the naming, you should follow the symbols: a_0 is an element of `E` which is coerced to `WithTop E` in this expression.

Reply:
Roger.